File: __tests__/globals_test.ml
open Jest

let () =
  test "pass" (fun () ->
    pass);

  Skip.test "fail" (fun () ->
    fail "");


  test "test" (fun () ->
    pass);
    
  Skip.test "test - expect fail" (fun () ->
    fail "");
  

  testAsync "testAsync" (fun finish ->
    finish pass);
    
  Skip.testAsync "testAsync - no done" (fun _ -> ());

  Skip.testAsync "testAsync - expect fail" (fun finish ->
    finish (fail ""));

  testAsync "testAsync - timeout ok" ~timeout:1 (fun finish ->
    finish pass);
    
  Skip.testAsync "testAsync - timeout fail" ~timeout:1 (fun _ -> ());

  
  testPromise "testPromise" (fun () ->
    Js.Promise.resolve pass);

  Skip.testPromise "testPromise - reject" (fun () ->
    Js.Promise.reject (Failure ""));

  Skip.testPromise "testPromise - expect fail" (fun () ->
    Js.Promise.resolve (fail ""));

  testPromise "testPromise - timeout ok" ~timeout:1 (fun () ->
    Js.Promise.resolve pass);
    
  Skip.testPromise "testPromise - timeout fail" ~timeout:1 (fun () ->
    Js.Promise.make (fun ~resolve:_ ~reject:_ -> ()));


  testAll "testAll" ["foo"; "bar"; "baz"] (fun input ->
    if Js.String.length input == 3 then pass else fail "");
  testAll "testAll - tuples" [("foo", 3); ("barbaz", 6); ("bananas!", 8)] (fun (input, output) ->
    if Js.String.length input == output then pass else fail "");
  

  describe "describe" (fun () ->
    test "some aspect" (fun () -> pass)
  );
  

  describe "beforeAll" (fun () -> 
    let x = ref 0 in
    
    beforeAll (fun () -> x := !x + 4);

    test "x is 4" (fun () -> if !x == 4 then pass else fail "");
    test "x is still 4" (fun () -> if !x == 4 then pass else fail "");
  );
  
  describe "beforeAllAsync" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      beforeAllAsync (fun (finish) -> x := !x + 4; finish ());

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is still 4" (fun () -> if !x == 4 then pass else fail "");
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      beforeAllAsync ~timeout:100 (fun (finish) -> x := !x + 4; finish ());

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is still 4" (fun () -> if !x == 4 then pass else fail "");
    );

    Skip.describe "timeout should fail suite" (fun () ->
      (* This apparently runs even if the suite is skipped *)
      (*beforeAllAsync ~timeout:1 (fun _ ->());*)

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );

  describe "beforeAllPromise" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      beforeAllPromise (fun () -> x := !x + 4; Js.Promise.resolve ());

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is still 4" (fun () -> if !x == 4 then pass else fail "");
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      beforeAllPromise ~timeout:100 (fun () -> x := !x + 4; Js.Promise.resolve ());

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is still 4" (fun () -> if !x == 4 then pass else fail "");
    );

    Skip.describe "timeout should fail suite" (fun () ->
      (* This apparently runs even if the suite is skipped *)
      (*beforeAllPromise ~timeout:1 (fun () -> Js.Promise.make (fun ~resolve:_ ~reject:_ -> ()));*)

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );


  describe "beforeEach" (fun () -> 
    let x = ref 0 in
    
    beforeEach (fun () -> x := !x + 4);

    test "x is 4" (fun () -> if !x == 4 then pass else fail "");
    test "x is suddenly 8" (fun () -> if !x == 8 then pass else fail "");
  );
 
  describe "beforeEachAsync" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      beforeEachAsync (fun (finish) -> x := !x + 4; finish ());

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is suddenly 8" (fun () -> if !x == 8 then pass else fail "");
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      beforeEachAsync ~timeout:100 (fun (finish) -> x := !x + 4; finish ());

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is suddenly 8" (fun () -> if !x == 8 then pass else fail "");
    );

    Skip.describe "timeout should fail suite" (fun () ->
      beforeEachAsync ~timeout:1 (fun _ -> ());

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );

  describe "beforeEachPromise" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      beforeEachPromise (fun () -> x := !x + 4; Js.Promise.resolve true);

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is suddenly 8" (fun () -> if !x == 8 then pass else fail "");
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      beforeEachPromise ~timeout:100 (fun () -> x := !x + 4; Js.Promise.resolve true);

      test "x is 4" (fun () -> if !x == 4 then pass else fail "");
      test "x is suddenly 8" (fun () -> if !x == 8 then pass else fail "");
    );

    Skip.describe "timeout should fail suite" (fun () ->
      beforeEachPromise ~timeout:1 (fun () -> Js.Promise.make (fun ~resolve:_ ~reject:_ -> ()));

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  ); 


  describe "afterAll" (fun () -> 
    let x = ref 0 in
    
    describe "phase 1" (fun () ->
      afterAll (fun () -> x := !x + 4);

      test "x is 0" (fun () -> if !x == 0 then pass else fail "");
      test "x is still 0" (fun () -> if !x == 0 then pass else fail "");
    );
    
    describe "phase 2" (fun () -> 
      test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
    );
  );
  
  describe "afterAllAsync" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      describe "phase 1" (fun () ->
        afterAllAsync (fun (finish) -> x := !x + 4; finish ());

        test "x is 0" (fun () -> if !x == 0 then pass else fail "");
        test "x is still 0" (fun () -> if !x == 0 then pass else fail "");
      );

      describe "phase 2" (fun () ->
        test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
      );
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      describe "phase 1" (fun () ->
        afterAllAsync ~timeout:100 (fun (finish) -> x := !x + 4; finish ());

        test "x is 0" (fun () -> if !x == 0 then pass else fail "");
        test "x is still 0" (fun () -> if !x == 0 then pass else fail "");
      );

      describe "phase 2" (fun () ->
        test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
      );
    );

    describe "timeout should not fail suite" (fun () ->
      afterAllAsync ~timeout:1 (fun _ ->());

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );

  describe "afterAllPromise" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      describe "phase 1" (fun () ->
        afterAllPromise (fun () -> x := !x + 4; Js.Promise.resolve true);

        test "x is 0" (fun () -> if !x == 0 then pass else fail "");
        test "x is still 0" (fun () -> if !x == 0 then pass else fail "");
      );

      describe "phase 2" (fun () ->
        test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
      );
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      describe "phase 1" (fun () ->
        afterAllPromise ~timeout:100 (fun () -> x := !x + 4; Js.Promise.resolve true);

        test "x is 0" (fun () -> if !x == 0 then pass else fail "");
        test "x is still 0" (fun () -> if !x == 0 then pass else fail "");
      );

      describe "phase 2" (fun () ->
        test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
      );
    );

    describe "timeout should not fail suite" (fun () ->
      afterAllPromise ~timeout:1 (fun () -> Js.Promise.make (fun ~resolve:_ ~reject:_ -> ()));

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );


  describe "afterEach" (fun () -> 
    let x = ref 0 in
    
    afterEach (fun () -> x := !x + 4);

    test "x is 0" (fun () -> if !x == 0 then pass else fail "");
    test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
  );
  
  describe "afterEachAsync" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      afterEachAsync (fun (finish) -> x := !x + 4; finish ());

      test "x is 0" (fun () -> if !x == 0 then pass else fail "");
      test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      afterEachAsync ~timeout:100 (fun (finish) -> x := !x + 4; finish ());

      test "x is 0" (fun () -> if !x == 0 then pass else fail "");
      test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
    );

    Skip.describe "timeout should fail suite" (fun () ->
      afterEachAsync ~timeout:1 (fun _ -> ());

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );

  describe "afterEachPromise" (fun () ->
    describe "without timeout" (fun () ->
      let x = ref 0 in

      afterEachPromise (fun () -> x := !x + 4; Js.Promise.resolve true);

      test "x is 0" (fun () -> if !x == 0 then pass else fail "");
      test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
    );

    describe "with 100ms timeout" (fun () ->
      let x = ref 0 in

      afterEachPromise ~timeout:100 (fun () -> x := !x + 4; Js.Promise.resolve true);

      test "x is 0" (fun () -> if !x == 0 then pass else fail "");
      test "x is suddenly 4" (fun () -> if !x == 4 then pass else fail "");
    );

    Skip.describe "timeout should fail suite" (fun () ->
      afterEachPromise ~timeout:1 (fun () -> Js.Promise.make (fun ~resolve:_ ~reject:_ -> ()));

      test "" (fun () -> pass); (* runner will crash if there's no tests *)
    );
  );


  describe "Only" (fun () ->
   (* See globals_only_test.ml *)
   ()
  );


  describe "Skip" (fun () ->
    Skip.test "Skip.test" (fun () -> pass);

    Skip.testAsync "Skip.testAsync" (fun finish ->
      finish pass);
    Skip.testAsync "Skip.testAsync - timeout" ~timeout:1 (fun _ -> ());

    Skip.testPromise "Skip.testPromise" (fun () ->
      Js.Promise.resolve pass);
    Skip.testPromise "testPromise - timeout" ~timeout:1 (fun () ->
      Js.Promise.make (fun ~resolve:_ ~reject:_ -> ()));

    Skip.testAll "testAll" ["foo"; "bar"; "baz"] (fun input ->
      if Js.String.length input == 3 then pass else fail "");
    Skip.testAll "testAll - tuples" [("foo", 3); ("barbaz", 6); ("bananas!", 8)] (fun (input, output) ->
      if Js.String.length input == output then pass else fail "");

    Skip.describe "Skip.describe" (fun () ->
      test "some aspect" (fun () -> pass);
    );

  );
  
  describe "Todo" (fun () ->
    Todo.test "Todo.test";
  )

File: lib/js/__tests__/globals_test.js
'use strict';

var Jest                    = require("../src/jest.js");
var Curry                   = require("bs-platform/lib/js/curry.js");
var Caml_builtin_exceptions = require("bs-platform/lib/js/caml_builtin_exceptions.js");

Jest.test("pass", (function () {
        return Jest.pass;
      }));

it.skip("fail", (function () {
        return Jest.fail("");
      }));

Jest.test("test", (function () {
        return Jest.pass;
      }));

it.skip("test - expect fail", (function () {
        return Jest.fail("");
      }));

Jest.testAsync("testAsync", (function (finish) {
        return Curry._1(finish, Jest.pass);
      }));

it.skip("testAsync - no done", (function () {
        return /* () */0;
      }));

it.skip("testAsync - expect fail", (function (finish) {
        return Curry._1(finish, Jest.fail(""));
      }));

Jest.testPromise("testPromise", (function () {
        return Promise.resolve(Jest.pass);
      }));

it.skip("testPromise - reject", (function () {
        return Promise.reject([
                    Caml_builtin_exceptions.failure,
                    ""
                  ]);
      }));

it.skip("testPromise - expect fail", (function () {
        return Promise.resolve(Jest.fail(""));
      }));

Jest.testAll("testAll", /* :: */[
      "foo",
      /* :: */[
        "bar",
        /* :: */[
          "baz",
          /* [] */0
        ]
      ]
    ], (function (input) {
        if (input.length === 3) {
          return Jest.pass;
        } else {
          return Jest.fail("");
        }
      }));

Jest.testAll("testAll - tuples", /* :: */[
      /* tuple */[
        "foo",
        3
      ],
      /* :: */[
        /* tuple */[
          "barbaz",
          6
        ],
        /* :: */[
          /* tuple */[
            "bananas!",
            8
          ],
          /* [] */0
        ]
      ]
    ], (function (param) {
        if (param[0].length === param[1]) {
          return Jest.pass;
        } else {
          return Jest.fail("");
        }
      }));

describe("describe", (function () {
        return Jest.test("some aspect", (function () {
                      return Jest.pass;
                    }));
      }));

describe("beforeAll", (function () {
        var x = [0];
        beforeAll((function () {
                x[0] = x[0] + 4 | 0;
                return /* () */0;
              }));
        Jest.test("x is 4", (function () {
                if (x[0] === 4) {
                  return Jest.pass;
                } else {
                  return Jest.fail("");
                }
              }));
        return Jest.test("x is still 4", (function () {
                      if (x[0] === 4) {
                        return Jest.pass;
                      } else {
                        return Jest.fail("");
                      }
                    }));
      }));

describe("beforeAllAsync", (function () {
        describe("without timeout", (function () {
                var x = [0];
                Jest.beforeAllAsync(/* None */0, (function (finish) {
                        x[0] = x[0] + 4 | 0;
                        return Curry._1(finish, /* () */0);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is still 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                Jest.beforeAllAsync(/* Some */[100], (function (finish) {
                        x[0] = x[0] + 4 | 0;
                        return Curry._1(finish, /* () */0);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is still 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe.skip("timeout should fail suite", (function () {
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("beforeAllPromise", (function () {
        describe("without timeout", (function () {
                var x = [0];
                Jest.beforeAllPromise(/* None */0, (function () {
                        x[0] = x[0] + 4 | 0;
                        return Promise.resolve(/* () */0);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is still 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                Jest.beforeAllPromise(/* Some */[100], (function () {
                        x[0] = x[0] + 4 | 0;
                        return Promise.resolve(/* () */0);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is still 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe.skip("timeout should fail suite", (function () {
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("beforeEach", (function () {
        var x = [0];
        beforeEach((function () {
                x[0] = x[0] + 4 | 0;
                return /* () */0;
              }));
        Jest.test("x is 4", (function () {
                if (x[0] === 4) {
                  return Jest.pass;
                } else {
                  return Jest.fail("");
                }
              }));
        return Jest.test("x is suddenly 8", (function () {
                      if (x[0] === 8) {
                        return Jest.pass;
                      } else {
                        return Jest.fail("");
                      }
                    }));
      }));

describe("beforeEachAsync", (function () {
        describe("without timeout", (function () {
                var x = [0];
                Jest.beforeEachAsync(/* None */0, (function (finish) {
                        x[0] = x[0] + 4 | 0;
                        return Curry._1(finish, /* () */0);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is suddenly 8", (function () {
                              if (x[0] === 8) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                Jest.beforeEachAsync(/* Some */[100], (function (finish) {
                        x[0] = x[0] + 4 | 0;
                        return Curry._1(finish, /* () */0);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is suddenly 8", (function () {
                              if (x[0] === 8) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe.skip("timeout should fail suite", (function () {
                Jest.beforeEachAsync(/* Some */[1], (function () {
                        return /* () */0;
                      }));
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("beforeEachPromise", (function () {
        describe("without timeout", (function () {
                var x = [0];
                Jest.beforeEachPromise(/* None */0, (function () {
                        x[0] = x[0] + 4 | 0;
                        return Promise.resolve(/* true */1);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is suddenly 8", (function () {
                              if (x[0] === 8) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                Jest.beforeEachPromise(/* Some */[100], (function () {
                        x[0] = x[0] + 4 | 0;
                        return Promise.resolve(/* true */1);
                      }));
                Jest.test("x is 4", (function () {
                        if (x[0] === 4) {
                          return Jest.pass;
                        } else {
                          return Jest.fail("");
                        }
                      }));
                return Jest.test("x is suddenly 8", (function () {
                              if (x[0] === 8) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe.skip("timeout should fail suite", (function () {
                Jest.beforeEachPromise(/* Some */[1], (function () {
                        return new Promise((function (_, _$1) {
                                      return /* () */0;
                                    }));
                      }));
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("afterAll", (function () {
        var x = [0];
        describe("phase 1", (function () {
                afterAll((function () {
                        x[0] = x[0] + 4 | 0;
                        return /* () */0;
                      }));
                Jest.test("x is 0", (function () {
                        if (x[0]) {
                          return Jest.fail("");
                        } else {
                          return Jest.pass;
                        }
                      }));
                return Jest.test("x is still 0", (function () {
                              if (x[0]) {
                                return Jest.fail("");
                              } else {
                                return Jest.pass;
                              }
                            }));
              }));
        describe("phase 2", (function () {
                return Jest.test("x is suddenly 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        return /* () */0;
      }));

describe("afterAllAsync", (function () {
        describe("without timeout", (function () {
                var x = [0];
                describe("phase 1", (function () {
                        Jest.afterAllAsync(/* None */0, (function (finish) {
                                x[0] = x[0] + 4 | 0;
                                return Curry._1(finish, /* () */0);
                              }));
                        Jest.test("x is 0", (function () {
                                if (x[0]) {
                                  return Jest.fail("");
                                } else {
                                  return Jest.pass;
                                }
                              }));
                        return Jest.test("x is still 0", (function () {
                                      if (x[0]) {
                                        return Jest.fail("");
                                      } else {
                                        return Jest.pass;
                                      }
                                    }));
                      }));
                describe("phase 2", (function () {
                        return Jest.test("x is suddenly 4", (function () {
                                      if (x[0] === 4) {
                                        return Jest.pass;
                                      } else {
                                        return Jest.fail("");
                                      }
                                    }));
                      }));
                return /* () */0;
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                describe("phase 1", (function () {
                        Jest.afterAllAsync(/* Some */[100], (function (finish) {
                                x[0] = x[0] + 4 | 0;
                                return Curry._1(finish, /* () */0);
                              }));
                        Jest.test("x is 0", (function () {
                                if (x[0]) {
                                  return Jest.fail("");
                                } else {
                                  return Jest.pass;
                                }
                              }));
                        return Jest.test("x is still 0", (function () {
                                      if (x[0]) {
                                        return Jest.fail("");
                                      } else {
                                        return Jest.pass;
                                      }
                                    }));
                      }));
                describe("phase 2", (function () {
                        return Jest.test("x is suddenly 4", (function () {
                                      if (x[0] === 4) {
                                        return Jest.pass;
                                      } else {
                                        return Jest.fail("");
                                      }
                                    }));
                      }));
                return /* () */0;
              }));
        describe("timeout should not fail suite", (function () {
                Jest.afterAllAsync(/* Some */[1], (function () {
                        return /* () */0;
                      }));
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("afterAllPromise", (function () {
        describe("without timeout", (function () {
                var x = [0];
                describe("phase 1", (function () {
                        Jest.afterAllPromise(/* None */0, (function () {
                                x[0] = x[0] + 4 | 0;
                                return Promise.resolve(/* true */1);
                              }));
                        Jest.test("x is 0", (function () {
                                if (x[0]) {
                                  return Jest.fail("");
                                } else {
                                  return Jest.pass;
                                }
                              }));
                        return Jest.test("x is still 0", (function () {
                                      if (x[0]) {
                                        return Jest.fail("");
                                      } else {
                                        return Jest.pass;
                                      }
                                    }));
                      }));
                describe("phase 2", (function () {
                        return Jest.test("x is suddenly 4", (function () {
                                      if (x[0] === 4) {
                                        return Jest.pass;
                                      } else {
                                        return Jest.fail("");
                                      }
                                    }));
                      }));
                return /* () */0;
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                describe("phase 1", (function () {
                        Jest.afterAllPromise(/* Some */[100], (function () {
                                x[0] = x[0] + 4 | 0;
                                return Promise.resolve(/* true */1);
                              }));
                        Jest.test("x is 0", (function () {
                                if (x[0]) {
                                  return Jest.fail("");
                                } else {
                                  return Jest.pass;
                                }
                              }));
                        return Jest.test("x is still 0", (function () {
                                      if (x[0]) {
                                        return Jest.fail("");
                                      } else {
                                        return Jest.pass;
                                      }
                                    }));
                      }));
                describe("phase 2", (function () {
                        return Jest.test("x is suddenly 4", (function () {
                                      if (x[0] === 4) {
                                        return Jest.pass;
                                      } else {
                                        return Jest.fail("");
                                      }
                                    }));
                      }));
                return /* () */0;
              }));
        describe("timeout should not fail suite", (function () {
                Jest.afterAllPromise(/* Some */[1], (function () {
                        return new Promise((function (_, _$1) {
                                      return /* () */0;
                                    }));
                      }));
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("afterEach", (function () {
        var x = [0];
        afterEach((function () {
                x[0] = x[0] + 4 | 0;
                return /* () */0;
              }));
        Jest.test("x is 0", (function () {
                if (x[0]) {
                  return Jest.fail("");
                } else {
                  return Jest.pass;
                }
              }));
        return Jest.test("x is suddenly 4", (function () {
                      if (x[0] === 4) {
                        return Jest.pass;
                      } else {
                        return Jest.fail("");
                      }
                    }));
      }));

describe("afterEachAsync", (function () {
        describe("without timeout", (function () {
                var x = [0];
                Jest.afterEachAsync(/* None */0, (function (finish) {
                        x[0] = x[0] + 4 | 0;
                        return Curry._1(finish, /* () */0);
                      }));
                Jest.test("x is 0", (function () {
                        if (x[0]) {
                          return Jest.fail("");
                        } else {
                          return Jest.pass;
                        }
                      }));
                return Jest.test("x is suddenly 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                Jest.afterEachAsync(/* Some */[100], (function (finish) {
                        x[0] = x[0] + 4 | 0;
                        return Curry._1(finish, /* () */0);
                      }));
                Jest.test("x is 0", (function () {
                        if (x[0]) {
                          return Jest.fail("");
                        } else {
                          return Jest.pass;
                        }
                      }));
                return Jest.test("x is suddenly 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe.skip("timeout should fail suite", (function () {
                Jest.afterEachAsync(/* Some */[1], (function () {
                        return /* () */0;
                      }));
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("afterEachPromise", (function () {
        describe("without timeout", (function () {
                var x = [0];
                Jest.afterEachPromise(/* None */0, (function () {
                        x[0] = x[0] + 4 | 0;
                        return Promise.resolve(/* true */1);
                      }));
                Jest.test("x is 0", (function () {
                        if (x[0]) {
                          return Jest.fail("");
                        } else {
                          return Jest.pass;
                        }
                      }));
                return Jest.test("x is suddenly 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe("with 100ms timeout", (function () {
                var x = [0];
                Jest.afterEachPromise(/* Some */[100], (function () {
                        x[0] = x[0] + 4 | 0;
                        return Promise.resolve(/* true */1);
                      }));
                Jest.test("x is 0", (function () {
                        if (x[0]) {
                          return Jest.fail("");
                        } else {
                          return Jest.pass;
                        }
                      }));
                return Jest.test("x is suddenly 4", (function () {
                              if (x[0] === 4) {
                                return Jest.pass;
                              } else {
                                return Jest.fail("");
                              }
                            }));
              }));
        describe.skip("timeout should fail suite", (function () {
                Jest.afterEachPromise(/* Some */[1], (function () {
                        return new Promise((function (_, _$1) {
                                      return /* () */0;
                                    }));
                      }));
                return Jest.test("", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

describe("Only", (function () {
        return /* () */0;
      }));

describe("Skip", (function () {
        it.skip("Skip.test", (function () {
                return Jest.pass;
              }));
        it.skip("Skip.testAsync", (function (finish) {
                return Curry._1(finish, Jest.pass);
              }));
        it.skip("Skip.testPromise", (function () {
                return Promise.resolve(Jest.pass);
              }));
        it.skip("testAll", /* :: */[
              "foo",
              /* :: */[
                "bar",
                /* :: */[
                  "baz",
                  /* [] */0
                ]
              ]
            ], (function (input) {
                if (input.length === 3) {
                  return Jest.pass;
                } else {
                  return Jest.fail("");
                }
              }));
        it.skip("testAll - tuples", /* :: */[
              /* tuple */[
                "foo",
                3
              ],
              /* :: */[
                /* tuple */[
                  "barbaz",
                  6
                ],
                /* :: */[
                  /* tuple */[
                    "bananas!",
                    8
                  ],
                  /* [] */0
                ]
              ]
            ], (function (param) {
                if (param[0].length === param[1]) {
                  return Jest.pass;
                } else {
                  return Jest.fail("");
                }
              }));
        describe.skip("Skip.describe", (function () {
                return Jest.test("some aspect", (function () {
                              return Jest.pass;
                            }));
              }));
        return /* () */0;
      }));

/*  Not a pure module */
